pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,207 @@
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "eventqueue")
+)
+
+// EventQueue is a structure which is utilized to handle Events in a first-in,
+// first-out order.There are several important properties for an EventQueue. An
+// EventQueue may be implemented in a blocking manner, such that only one Event
+// is able to be consumed at a time per event queue. This is useful in the case
+// for ensuring that certain types of events do not run concurrently. An
+// EventQueue may be closed, in which case all events which are queued up, but
+// have not been processed yet, will be cancelled (i.e., not ran). It is
+// guaranteed that no events will be scheduled onto an EventQueue after it has
+// been closed; if any event is attempted to be scheduled onto an EventQueue
+// after it has been closed, it will be cancelled immediately. For any event to
+// be processed by the EventQueue, it must implement the `EventHandler`
+// interface. This allows for different types of events to be processed by
+// anything which chooses to utilize an `EventQueue`.
+type EventQueue struct {
+	// This should always be a buffered channel.
+	events chan *Event
+	// close is closed once the EventQueue has been closed.
+	close chan struct{}
+	// drained is closed once the events channel is drained.
+	drained chan struct{}
+	// eventQueueOnce is used to ensure that the EventQueue business logic can
+	// only be ran once.
+	eventQueueOnce sync.Once
+}
+
+// NewEventQueue returns an EventQueue with a capacity for only one event at
+// a time, and all other needed fields initialized.
+func NewEventQueue() *EventQueue {
+	return &EventQueue{
+		// Only one event can be consumed at a time.
+		events:  make(chan *Event, 1),
+		close:   make(chan struct{}),
+		drained: make(chan struct{}),
+	}
+
+}
+
+// Event is an event that can be enqueued onto an EventQueue.
+type Event struct {
+	// Metadata is the information about the event which is sent
+	// by its queuer. Metadata must implement the EventHandler interface in
+	// order for the Event to be successfully processed by the EventQueue.
+	Metadata interface{}
+
+	// EventResults is a channel on which the results of the event are sent.
+	// It is populated by the EventQueue itself, not by the queuer.
+	EventResults chan interface{}
+
+	// Cancelled signals that the given Event was not ran. This can happen
+	// if the EventQueue processing this Event was closed before the Event was
+	// Enqueued onto the Event queue, or if the Event was Enqueued onto an
+	// EventQueue, and the EventQueue on which the Event was scheduled was
+	// closed.
+	Cancelled chan struct{}
+}
+
+// NewEvent returns an Event with all fields initialized.
+func NewEvent(meta interface{}) *Event {
+	return &Event{
+		Metadata:     meta,
+		EventResults: make(chan interface{}, 1),
+		Cancelled:    make(chan struct{}),
+	}
+}
+
+// WasCancelled returns whether the Cancelled channel for the given Event has
```

Comment:
Would be helpful if the description here says what cancellation means. Is the channel closed when the queuer cancels the event or when it times out?

Reply:
Will include an update in a new upload about the semantics of cancellation in the description of this function. 

Currently, there is not a capability of the queuer to cancel an event. The EventQueue is the manager of Event state once an Event is Enqueued.